robot-core/src/main/java/org/obolibrary/robot/QuotedEntityChecker.java

```diff
@@ -389,10 +389,15 @@ public OWLDataProperty getOWLDataProperty(@Nonnull String name) {
     if (iri != null) {
       return dataFactory.getOWLDataProperty(iri);
     }
-    if (ioHelper != null) {
-      iri = ioHelper.createIRI(name);
-      if (iri != null) {
-        return dataFactory.getOWLDataProperty(iri);
+    // prevent punning
+    if (!objectProperties.containsKey(name)) {
+      if (ioHelper != null) {
+        iri = ioHelper.createIRI(name);
+        if (iri != null) {
+          OWLDataProperty owlDataProperty = dataFactory.getOWLDataProperty(iri);
+          dataProperties.put(name, iri);
+          return owlDataProperty;
+        }
```

Comment:
Essential change?

Reply:
Yes, when if controls are deleted, owlapi is throwing [Illegal Punning exception](https://github.com/owlcs/owlapi/blob/owlapi-parent-4.5.22/parsers/src/main/java/org/semanticweb/owlapi/manchestersyntax/parser/ManchesterOWLSyntaxParserImpl.java#L703) in the following tests:
```
testChecker(org.obolibrary.robot.QuotedEntityCheckerTest): Encountered GO:XXXX at line 1 column 1. Expected one of:(..)
testLegacyDocsTemplate(org.obolibrary.robot.TemplateTest): template#MANCHESTER PARSE ERROR the expression (NCBITaxon:10116) at row 3, column 9 in table "/docs-template.csv" cannot be parsed: Encountered NCBITaxon:10116 at line 1 column 1. Expected one of:(..)
```

```
org.semanticweb.owlapi.manchestersyntax.renderer.ParserException: Encountered GO:XXXX at line 1 column 1. Expected one of:

	at org.semanticweb.owlapi.manchestersyntax.parser.ManchesterOWLSyntaxParserImpl$ExceptionBuilder.build(ManchesterOWLSyntaxParserImpl.java:2587)
	at org.semanticweb.owlapi.manchestersyntax.parser.ManchesterOWLSyntaxParserImpl.parseNonNaryClassExpression(ManchesterOWLSyntaxParserImpl.java:705)
	at org.semanticweb.owlapi.manchestersyntax.parser.ManchesterOWLSyntaxParserImpl.parseIntersection(ManchesterOWLSyntaxParserImpl.java:610)
	at org.semanticweb.owlapi.manchestersyntax.parser.ManchesterOWLSyntaxParserImpl.parseUnion(ManchesterOWLSyntaxParserImpl.java:633)
	at org.semanticweb.owlapi.manchestersyntax.parser.ManchesterOWLSyntaxParserImpl.parseClassExpression(ManchesterOWLSyntaxParserImpl.java:592)
	at org.semanticweb.owlapi.manchestersyntax.parser.ManchesterOWLSyntaxClassExpressionParser.parse(ManchesterOWLSyntaxClassExpressionParser.java:61)
	at org.obolibrary.robot.QuotedEntityCheckerTest.testChecker(QuotedEntityCheckerTest.java:85)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:542)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:770)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

Reply:
https://github.com/owlcs/owlapi/issues/851  
> Property punning between properties is illegal, but punning of a property name and class name or individual name is allowed.

Reply:
Yes, this is only related with the ROBOT template. This is effecting manchester syntax parsing when an input ontology is not provided and need to expand curies from the given jsonld context (`--prefixes`).

Reply:
So what would happen now if the `--input ontology.owl` contains `RO:123 rdf:type AP` and `RO:123 rdf:type OP` and the template contained this string: `AI '123'`? Would it fail to generate the annotation assertion?

Reply:
If `--input ontology.owl` is provided and referred term is in the ontology (no matter its type), everything will function as previous (the newly added if block is unreachable and not executed). Also, the new algorithm of the owlapi is not affecting annotation properties, it is only preventing punning of data and object properties. I applied the same logic to ROBOT.  [ManchesterOWLSyntaxParserImpl.java](https://github.com/owlcs/owlapi/blob/owlapi-parent-4.5.22/parsers/src/main/java/org/semanticweb/owlapi/manchestersyntax/parser/ManchesterOWLSyntaxParserImpl.java#L701):
```java
        boolean isClassName = isClassName(tok);
        boolean isDataPropertyName = isDataPropertyName(tok);
        boolean isObjectPropertyName = isObjectPropertyName(tok);
        if (isClassName && (isDataPropertyName || isObjectPropertyName)) {
            if (isDataPropertyName && isObjectPropertyName) {
``` 
The only limitation that this new code block may cause is: In case a CURIE referred in the template is not in the ontology and it is expanded as OP, it cannot be expanded as DP afterwards (and vice versa). But couldn't find a real-life robot scenario that will trigger this behavior.

Reply:
That makes total sense and alleviates all my concerns, thanks for explaining!